tensordict/prototype/tensorclass.py

```diff
@@ -158,72 +158,60 @@ def __torch_function__(
-            if missing_params:
-                n_missing = len(missing_params)
-                raise TypeError(
-                    f"{self.__class__.__name__}.__init__() missing {n_missing} "
-                    f"required positional argument{'' if n_missing == 1 else 's'}: "
-                    f"""{", ".join(f"'{name}'" for name in missing_params)}"""
-                )
+    def wrapper(self, *args, batch_size, device=None, **kwargs):
+        for value, key in zip(args, self.__dataclass_fields__):
+            if key in kwargs:
+                raise ValueError(f"The key {key} is already set in kwargs")
+            kwargs[key] = value
+
+        for key, field in self.__dataclass_fields__.items():
+            if field.default_factory is not dataclasses.MISSING:
+                default = field.default_factory()
+            else:
+                default = field.default
+            if default not in (None, dataclasses.MISSING):
+                kwargs.setdefault(key, default)
 
-            self.tensordict = TensorDict({}, batch_size=batch_size, device=device)
-            init(
-                self, **{key: _get_typed_value(value) for key, value in kwargs.items()}
+        missing_params = [p for p in required_params if p not in kwargs]
+        if missing_params:
+            n_missing = len(missing_params)
+            raise TypeError(
+                f"{self.__class__.__name__}.__init__() missing {n_missing} "
+                f"required positional argument{'' if n_missing == 1 else 's'}: "
+                f"""{", ".join(f"'{name}'" for name in missing_params)}"""
             )
 
+        self.tensordict = TensorDict({}, batch_size=batch_size, device=device)
+        init(self, **{key: _get_typed_value(value) for key, value in kwargs.items()})
+
     new_params = [
-        inspect.Parameter(
-            "batch_size", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
-        inspect.Parameter(
-            "device", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
-        inspect.Parameter(
-            "_tensordict", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
+        inspect.Parameter("batch_size", inspect.Parameter.KEYWORD_ONLY),
+        inspect.Parameter("device", inspect.Parameter.KEYWORD_ONLY, default=None),
     ]
     wrapper.__signature__ = init_sig.replace(parameters=params + new_params)
 
     return wrapper
 
 
-def _build_from_tensordict(cls, tensordict):
-    return cls(_tensordict=tensordict)
+def _from_tensordict_wrapper(expected_keys):
+    def wrapper(cls, tensordict):
+        if not all(key in expected_keys for key in tensordict.keys()):
+            raise ValueError(
+                f"Keys from the tensordict ({set(tensordict.keys())}) must "
+                f"correspond to the class attributes ({expected_keys})."
+            )
+        tc = cls(**tensordict, batch_size=tensordict.batch_size)
```

Comment:
isn't this a bit overkilling it?
IIUC we deconstruct the tensordict, reconstruct another one, then replace it with the old one. Since the construction is quite time consuming, do you think we can find another solution?

Reply:
my very same objection, but @tcbegley measured the overhead and seems negligible. Do you have a specific test case in mind that we could try to double check the benchmark?

Reply:
The overhead is mainly due to the checks, so if we could do the checks only once it'd be cool (`TensorDIct(dictionary, batch_size, device, _run_checks=False)`).
Given that the tensordict that is provided is presumably already checked we should be good no?

Reply:
off-topic: why the `_` in a parameter name? It should be used only for protected methods/variables

Reply:
Exactly for that: we don't want users to call that. We may refactor this in the future to hide it a bit more.
You should use that only if you build a tensordict out of another tensordict and the checks have been done, which will only occur in with dedicated methods.
I'm open to suggestions to make it cleaner :)

Reply:
See this example:
```python
td = TensorDict({"a": torch.randn(3), "b": torch.randn(2)}, [3], _run_checks=False) # passes 
td[2]  # breaks
```
the code breaks where it should not.
If we make that arg public, we're telling the users that this would be ok-ish to construct the TD like this, which isn't.

Reply:
Since we initialise an empty tensordict in the constructor, and then populate it via `__setattr__` from inside the dataclass' constructor, I think we will also need to set `_run_checks=False` inside there to really reduce overhead. Is it ok to always have that turned off though? Presumably we still want checks when running something like

```python
tc.X = torch.rand(100, 10)
```

Reply:
> Since we initialise an empty tensordict in the constructor, and then populate it via `__setattr__` from inside the dataclass' constructor, I think we will also need to set `_run_checks=False` inside there to really reduce overhead. Is it ok to always have that turned off though? Presumably we still want checks when running something like
> 
> ```python
> tc.X = torch.rand(100, 10)
> ```

yep in that case we need to run it

Reply:
> Exactly for that: we don't want users to call that. We may refactor this in the future to hide it a bit more. You should use that only if you build a tensordict out of another tensordict and the checks have been done, which will only occur in with dedicated methods. I'm open to suggestions to make it cleaner :)

couldn't we simply store a `self._checked` boolean attribute that gets set to true the first time, and get rid of that argument constructor?

Reply:
That could be an option, I should go through the tensordict.py file and check if that would work in all cases